### ISSUE TITLE
Prediction for changes in meta inventories

### ISSUE THREAD

**Issue by MirceaKitsune** [minetest/minetest#1374]
This is a well known problem, but since it's still one of the most problematic issues in Minetest IMO, I decided to report it again in hopes something can be done about it.

When moving items inside your own inventory, the changes are always applied immediately. But when moving stuff in node inventories, there is a lot of lag. Even in a single player world, it takes about a second after you drag an item to a slot for that item to actually go there.

This causes confusion and also feels horrible... especially when moving multiple items of one type or splitting / combining stacks. With workbenches it's a nightmare, and it's also problematic when sorting your things inside chests.

Please add local prediction of inventory changes to meta inventories. So when you move an item in a chest for example, the visual change is instantly applied locally rather than waiting for the server to notify it. The list shall of course still be updated when the server sends changes (current behavior), but also predicted locally before that. This would completely eliminate perceived lag... except for cases where a change in the formspec depends on server actions, but this is understandable and a way smaller problem.

As I already discussed with sapier on IRC, a small obstacle are areas that the player is not allowed to place things in... such as the craft result / furnace product slot. I assume the best solution is having the client know in which inventory lists the player is allowed to drop items, which is part of the formspec definition so it shouldn't be an issue.

**Comment by hunter688** [comment]
Crafting, moving items between inventories is big part of game. I dont know why this issue does not have at least medium priority :-(

**Comment by bgh251f2** [comment]
This is the main thing stopping me and people I know from playing this game. I don't really understand why it is set to low priority. Can anyone explain?

**Comment by asl97** [comment]
@bgh251f2 the low-priority label has long been remove.

**Comment by bgh251f2** [comment]
@asl97 Yeah, my bad. I should have paid more attention. Shame it hasn't been fixed yet.

**Comment by rubenwardy** [comment]
bump, this may be worth discussing

**Comment by Zughy** [comment]
It doesn't seem to me that it lags anymore (see video down below). Is the issue still valid, lag aside?

https://user-images.githubusercontent.com/63455151/171832860-fa1dbd86-540b-4660-8db4-7ffec35eb04e.mp4

**Comment by SmallJoker** [comment]
https://github.com/minetest/minetest/blob/9fc018ded10225589d2559d24a5db739e891fb31/src/inventorymanager.cpp#L617-L621

No, this issue theoretically still exists. Only player inventory movements + take + drop events are predicted. However, the effect might not be easily encountered any more thanks to network improvements.

**Comment by erlehmann** [comment]
@SmallJoker I never encountered it in singleplayer unless the server was lagging so much the game was not playable anyway. In multiplayer it seems semi-common though. I think aside from network lag this can also happen when the server is just a bit busy.

A problem with the prediction might be that what is allowed could be context sensitive, but I am not sure how often that happens.
